Follow up on #2427, CMake problems

## Your Environment
* Operating System and version: Ubuntu 18.04
* Compiler: GCC 7
* PCL Version: Latest

## Context
This is the minimal example to recreate the problem

CMakeLists.txt
```
cmake_minimum_required(VERSION 3.11)

add_subdirectory(sample2)
add_subdirectory(sample3)
```

sample2/CMakeLists.txt
```
cmake_minimum_required(VERSION 3.11)

project(sample2)

find_package( PCL REQUIRED )

add_library(${PROJECT_NAME} SHARED Sample2.cpp Sample2.h)

target_link_libraries(${PROJECT_NAME} ${PCL_LIBRARIES})
```

sample2/Sample2.h
```
class Sample2{
 public:
  Sample2();
  ~Sample2();
};
```

sample2/Sample2.cpp
```
#include "Sample2.h"

Sample2::Sample2() {}

Sample2::~Sample2() {}
```

sample3/CMakeLists.txt
```
cmake_minimum_required(VERSION 3.11)

project(sample3)

find_package( PCL REQUIRED )

add_library(${PROJECT_NAME} SHARED Sample3.cpp Sample3.h)

target_link_libraries(${PROJECT_NAME} ${PCL_LIBRARIES})
```

sample3/Sample3.h
```
class Sample3{
 public:
  Sample3();
  ~Sample3();
};
```

sample3/Sample3.cpp
```
#include "Sample3.h"

#include "pcl/common/common.h"

Sample3::Sample3() {}

Sample3::~Sample3() {}
```

You can see that sample2 and sample3 are irrelevant to each other, but when I compile it shows the below compilation error

```
In file included from /usr/include/pcl-1.8/pcl/common/common.h:41:0,
                 from /home/hypevr/test/sample3/Sample3.cpp:3:
/usr/include/pcl-1.8/pcl/pcl_base.h:49:10: fatal error: Eigen/StdVector: No such file or directory
 #include <Eigen/StdVector>
          ^~~~~~~~~~~~~~~~~
compilation terminated.
sample3/CMakeFiles/sample3.dir/build.make:62: recipe for target 'sample3/CMakeFiles/sample3.dir/Sample3.cpp.o' failed
```

However, if I comment out the `add_subdirectory(sample2)` in the head CMakeLists so it becomes like this

```
cmake_minimum_required(VERSION 3.11)

#add_subdirectory(sample2)
add_subdirectory(sample3)
```

It actually compiles as it should be. So for some reason sample2 is still affecting sample3 even though they are irrelevant. Can someone please fix it?

--- a/PCLConfig.cmake.in
+++ b/PCLConfig.cmake.in
@@ -18,6 +18,7 @@ list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_LIST_DIR}/Modules")
 macro(pcl_report_not_found _reason)
   unset(PCL_FOUND)
   unset(PCL_LIBRARIES)
+  unset(PCL_COMPONENTS)
   unset(PCL_INCLUDE_DIRS)
   unset(PCL_LIBRARY_DIRS)
   unset(PCL_DEFINITIONS)
@@ -448,10 +449,13 @@ else(PCL_FIND_COMPONENTS)
   set(PCL_FIND_ALL 1)
 endif(PCL_FIND_COMPONENTS)
 
+compute_dependencies(PCL_TO_FIND_COMPONENTS)
+
 # We do not need to find components that have been found already, e.g. during previous invocation
 # of find_package(PCL). Filter them out.
 foreach(component ${PCL_TO_FIND_COMPONENTS})
-  if(NOT TARGET pcl_${component})
+  string(TOUPPER "${component}" COMPONENT)
+  if(NOT PCL_${COMPONENT}_FOUND)
     list(APPEND _PCL_TO_FIND_COMPONENTS ${component})
   endif()
 endforeach()
@@ -461,8 +465,6 @@ unset(_PCL_TO_FIND_COMPONENTS)
 if(NOT PCL_TO_FIND_COMPONENTS)
   return()
 endif()
-
-compute_dependencies(PCL_TO_FIND_COMPONENTS)
 
 # compute external dependencies per component
 foreach(component ${PCL_TO_FIND_COMPONENTS})
@@ -561,7 +563,7 @@ foreach(component ${PCL_TO_FIND_COMPONENTS})
       if(PCL_${COMPONENT}_LIBRARY_DEBUG)
         list(APPEND PCL_LIBRARY_DIRS ${component_library_path_debug})
       endif()
-      list(APPEND PCL_LIBRARIES ${pcl_component})
+      list(APPEND PCL_COMPONENTS ${pcl_component})
       mark_as_advanced(PCL_${COMPONENT}_LIBRARY PCL_${COMPONENT}_LIBRARY_DEBUG)
     endif()
     # Append internal dependencies
@@ -575,7 +577,7 @@ foreach(component ${PCL_TO_FIND_COMPONENTS})
       endif(PCL_${INT_DEP}_FOUND)
     endforeach(int_dep)
     if(_is_header_only EQUAL -1)
-      add_library(${pcl_component} @PCL_LIB_TYPE@ IMPORTED GLOBAL)
+      add_library(${pcl_component} @PCL_LIB_TYPE@ IMPORTED)
       if(PCL_${COMPONENT}_LIBRARY_DEBUG)
         set_target_properties(${pcl_component}
           PROPERTIES
@@ -621,8 +623,8 @@ if(NOT "${PCL_DEFINITIONS}" STREQUAL "")
   list(REMOVE_DUPLICATES PCL_DEFINITIONS)
 endif(NOT "${PCL_DEFINITIONS}" STREQUAL "")
 
-pcl_remove_duplicate_libraries(PCL_LIBRARIES PCL_DEDUP_LIBRARIES)
-set(PCL_LIBRARIES ${PCL_DEDUP_LIBRARIES})
+pcl_remove_duplicate_libraries(PCL_COMPONENTS PCL_LIBRARIES)
+
 # Add 3rd party libraries, as user code might include our .HPP implementations
 list(APPEND PCL_LIBRARIES ${BOOST_LIBRARIES} ${QHULL_LIBRARIES} ${OPENNI_LIBRARIES} ${OPENNI2_LIBRARIES} ${ENSENSO_LIBRARIES} ${davidSDK_LIBRARIES} ${DSSDK_LIBRARIES} ${RSSDK_LIBRARIES} ${FLANN_LIBRARIES} ${VTK_LIBRARIES})
 

